device.go

```diff
@@ -229,37 +231,174 @@ func virtioBlkDeviceHandler(_ context.Context, device pb.Device, spec *pb.Spec,
 	}
 
-	return updateSpecDeviceList(device, spec)
+	return updateSpecDevice(spec, devIdx, device.ContainerPath, device.VmPath, device.ContainerPath)
 }
 
 // device.Id should be the SCSI address of the disk in the format "scsiID:lunID"
-func virtioSCSIDeviceHandler(ctx context.Context, device pb.Device, spec *pb.Spec, s *sandbox) error {
+func virtioSCSIDeviceHandler(ctx context.Context, device pb.Device, spec *pb.Spec, s *sandbox, devIdx devIndex) error {
 	// Retrieve the device path from SCSI address.
 	devPath, err := getSCSIDevPath(s, device.Id)
 	if err != nil {
 		return err
 	}
 	device.VmPath = devPath
 
-	return updateSpecDeviceList(device, spec)
+	return updateSpecDevice(spec, devIdx, device.ContainerPath, device.VmPath, device.ContainerPath)
 }
 
-func nvdimmDeviceHandler(_ context.Context, device pb.Device, spec *pb.Spec, s *sandbox) error {
-	return updateSpecDeviceList(device, spec)
+func nvdimmDeviceHandler(_ context.Context, device pb.Device, spec *pb.Spec, s *sandbox, devIdx devIndex) error {
+	return updateSpecDevice(spec, devIdx, device.ContainerPath, device.VmPath, device.ContainerPath)
 }
 
-// updateSpecDeviceList takes a device description provided by the caller,
-// trying to find it on the guest. Once this device has been identified, the
-// "real" information that can be read from inside the VM is used to update
-// the same device in the list of devices provided through the OCI spec.
-// This is needed to update information about minor/major numbers that cannot
-// be predicted from the caller.
-func updateSpecDeviceList(device pb.Device, spec *pb.Spec) error {
-	// If no ContainerPath is provided, we won't be able to match and
-	// update the device in the OCI spec device list. This is an error.
-	if device.ContainerPath == "" {
+func vfioRebindDevice(bdf string) error {
+	fieldLogger := agentLog.WithField("guest-bdf", bdf)
+	fieldLogger.Debug("VFIO: Attempting to rebind")
+
+	sysPath := filepath.Join(sysBusPrefix, bdf)
+
+	// First we have to wait for the device to appear in the
+	// guest, since the hotplug can take a while.  We can't use
+	// the waiting logic in getDeviceName here, because the udev
+	// devices will only appear as we expect *after* we've
+	// rebound.
+	start := time.Now()
+	for {
+		_, err := os.Stat(sysPath)
+		if err == nil {
+			break
+		}
+		if !os.IsNotExist(err) {
+			return fmt.Errorf("Error accessing %s: %s", sysPath, err)
+		}
+		if elapsed := time.Since(start); elapsed > vfioTimeout {
+			return fmt.Errorf("Timeout waiting for %s (%v)", sysPath, elapsed)
+		}
+		time.Sleep(vfioHotplugPollInterval)
+	}
+	fieldLogger.Debugf("VFIO: Found PCI device after %v", time.Since(start))
+
+	overridePath := filepath.Join(sysPath, "driver_override")
+	err := ioutil.WriteFile(overridePath, []byte("vfio-pci"), 0200)
+	if err != nil {
+		return fmt.Errorf("Failed to write %s: %s", overridePath, err)
+	}
+
+	driverPath := filepath.Join(sysPath, "driver")
+	driver, err := os.Readlink(driverPath)
```

Comment:
This is reading the symbolic link, so I'd expect you would get something like '../../../bus/pci/drivers/xhci_hcd' at this stage (took that from a USB)

Reply:
Good point, I've added a line to chop this down to the base name.